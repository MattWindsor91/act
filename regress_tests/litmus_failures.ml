(* The Automagic Compiler Tormentor

   Copyright (c) 2018--2020 Matt Windsor and contributors

   ACT itself is licensed under the MIT License. See the LICENSE file in the
   project root for more information.

   ACT is based in part on code from the Herdtools7 project
   (https://github.com/herd/herdtools7) : see the LICENSE.herd file in the
   project root for more information. *)

(** Tests that the Litmus/C parser gives the appropriate error. *)

open Core

let fail_to_parse_file ~(file : Fpath.t) ~(path : Fpath.t) : unit Or_error.t
    =
  ignore file ;
  let result =
    Act_litmus_c.Frontend.Litmus.load (Plumbing.Input.of_fpath path)
  in
  Fmt.(pr "@[%a@]@." (result ~ok:(any "(success)") ~error:Error.pp)) result ;
  Ok ()

let run (test_dir : Fpath.t) : unit Or_error.t =
  let full_dir = Fpath.(test_dir / "litmus" / "failures" / "") in
  Common.regress_on_files "Litmus-failures" ~dir:full_dir ~ext:"litmus"
    ~f:fail_to_parse_file

let command =
  Common.make_regress_command ~summary:"checks parser failure messages" run

let () = Command.run command
