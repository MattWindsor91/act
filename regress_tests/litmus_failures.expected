# Litmus-failures tests

## cpp.c.litmus

```
("While reading from file:" input/litmus/failures/cpp.c.litmus
  "File \"input/litmus/failures/cpp.c.litmus\", line 13, characters 0-1:\
 \nlexing error: C preprocessor directives aren't supported.")
```
## location_fn.c.litmus

```
("While reading from file:" input/litmus/failures/location_fn.c.litmus
  "File \"input/litmus/failures/location_fn.c.litmus\", line 11, characters 9-10:\
 \nparse error: Expected a '{' to open a Litmus locations stanza.\
 \n(If this error occurs near a use of the word \"locations\" as an identifier,\
 \nrename it; it is clashing with a Litmus keyword.)\
 \n")
```
## tx_atomic_nonblock.litmus

```
("While reading from file:" input/litmus/failures/tx_atomic_nonblock.litmus
  "File \"input/litmus/failures/tx_atomic_nonblock.litmus\", line 14, characters 9-10:\
 \nparse error: Keyword `atomic` must be followed by a block\
 \n")
```
## tx_sync_nonblock.litmus

```
("While reading from file:" input/litmus/failures/tx_sync_nonblock.litmus
  "File \"input/litmus/failures/tx_sync_nonblock.litmus\", line 14, characters 15-16:\
 \nparse error: Keyword `synchronized` must be followed by a block\
 \n")
```

Ran 4 test(s).
